From 58a0fabeee969f25c1001720c801e92e8d4eb1d6 Mon Sep 17 00:00:00 2001
From: Zhong Jinghua <zhongjinghua@huawei.com>
Date: Mon, 10 Oct 2022 17:08:09 +0800
Subject: [PATCH 0596/1076] [Huawei] Revert "[Huawei] quota: Add more checking
 after reading from quota file"

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187046

--------------------------------

This reverts commit 0e2fdff7d55aa4ecb9f1e940216d1bd056b465a3.

This patch is temporary, revert it and backport the mainline patch.

Signed-off-by: Zhong Jinghua <zhongjinghua@huawei.com>
---
 fs/quota/quota_tree.c | 43 ++++++++++---------------------------------
 1 file changed, 10 insertions(+), 33 deletions(-)

diff --git a/fs/quota/quota_tree.c b/fs/quota/quota_tree.c
index 06e3ec528b2c..c200161554db 100644
--- a/fs/quota/quota_tree.c
+++ b/fs/quota/quota_tree.c
@@ -80,12 +80,12 @@ static ssize_t write_blk(struct qtree_mem_dqinfo *info, uint blk, char *buf)
 	return ret;
 }
 
-static inline int do_check_range(struct super_block *sb, const char *val_name,
-				 uint val, uint min_val, uint max_val)
+static inline int do_check_range(struct super_block *sb, uint val,
+				 uint min_val, uint max_val)
 {
 	if (val < min_val || val >= max_val) {
-		quota_error(sb, "Getting %s %u out of range %u-%u",
-			    val_name, val, min_val, max_val);
+		quota_error(sb, "Getting block %u out of range %u-%u",
+			    val, min_val, max_val);
 		return -EUCLEAN;
 	}
 
@@ -99,13 +99,11 @@ static int check_free_block(struct qtree_mem_dqinfo *info,
 	uint nextblk, prevblk;
 
 	nextblk = le32_to_cpu(dh->dqdh_next_free);
-	err = do_check_range(info->dqi_sb, "dqdh_next_free", nextblk, 0,
-			     info->dqi_blocks);
+	err = do_check_range(info->dqi_sb, nextblk, 0, info->dqi_blocks);
 	if (err)
 		return err;
 	prevblk = le32_to_cpu(dh->dqdh_prev_free);
-	err = do_check_range(info->dqi_sb, "dqdh_prev_free", prevblk, 0,
-			     info->dqi_blocks);
+	err = do_check_range(info->dqi_sb, prevblk, 0, info->dqi_blocks);
 	if (err)
 		return err;
 
@@ -279,11 +277,6 @@ static uint find_free_dqentry(struct qtree_mem_dqinfo *info,
 		*err = check_free_block(info, dh);
 		if (*err)
 			goto out_buf;
-		*err = do_check_range(info->dqi_sb, "dqdh_entries",
-				      le16_to_cpu(dh->dqdh_entries), 0,
-				      qtree_dqstr_in_blk(info));
-		if (*err)
-			goto out_buf;
 	} else {
 		blk = get_free_dqblk(info);
 		if ((int)blk < 0) {
@@ -365,10 +358,6 @@ static int do_insert_tree(struct qtree_mem_dqinfo *info, struct dquot *dquot,
 	}
 	ref = (__le32 *)buf;
 	newblk = le32_to_cpu(ref[get_index(info, dquot->dq_id, depth)]);
-	ret = do_check_range(dquot->dq_sb, "block", newblk, 0,
-			     info->dqi_blocks);
-	if (ret)
-		goto out_buf;
 	if (!newblk)
 		newson = 1;
 	if (depth == info->dqi_qtree_depth - 1) {
@@ -481,11 +470,6 @@ static int free_dqentry(struct qtree_mem_dqinfo *info, struct dquot *dquot,
 	}
 	dh = (struct qt_disk_dqdbheader *)buf;
 	ret = check_free_block(info, dh);
-	if (ret)
-		goto out_buf;
-	ret = do_check_range(info->dqi_sb, "dqdh_entries",
-			     le16_to_cpu(dh->dqdh_entries), 1,
-			     qtree_dqstr_in_blk(info) + 1);
 	if (ret)
 		goto out_buf;
 	le16_add_cpu(&dh->dqdh_entries, -1);
@@ -544,7 +528,7 @@ static int remove_tree(struct qtree_mem_dqinfo *info, struct dquot *dquot,
 		goto out_buf;
 	}
 	newblk = le32_to_cpu(ref[get_index(info, dquot->dq_id, depth)]);
-	ret = do_check_range(dquot->dq_sb, "block", newblk, QT_TREEOFF,
+	ret = do_check_range(dquot->dq_sb, newblk, QT_TREEOFF,
 			     info->dqi_blocks);
 	if (ret)
 		goto out_buf;
@@ -648,8 +632,7 @@ static loff_t find_tree_dqentry(struct qtree_mem_dqinfo *info,
 	blk = le32_to_cpu(ref[get_index(info, dquot->dq_id, depth)]);
 	if (!blk)	/* No reference? */
 		goto out_buf;
-	ret = do_check_range(dquot->dq_sb, "block", blk, QT_TREEOFF,
-			     info->dqi_blocks);
+	ret = do_check_range(dquot->dq_sb, blk, QT_TREEOFF, info->dqi_blocks);
 	if (ret)
 		goto out_buf;
 
@@ -765,13 +748,7 @@ static int find_next_id(struct qtree_mem_dqinfo *info, qid_t *id,
 		goto out_buf;
 	}
 	for (i = __get_index(info, *id, depth); i < epb; i++) {
-		uint blk_no = le32_to_cpu(ref[i]);
-
-		ret = do_check_range(info->dqi_sb, "block", blk_no, 0,
-				     info->dqi_blocks);
-		if (ret)
-			goto out_buf;
-		if (blk_no == 0) {
+		if (ref[i] == cpu_to_le32(0)) {
 			*id += level_inc;
 			continue;
 		}
@@ -779,7 +756,7 @@ static int find_next_id(struct qtree_mem_dqinfo *info, qid_t *id,
 			ret = 0;
 			goto out_buf;
 		}
-		ret = find_next_id(info, id, blk_no, depth + 1);
+		ret = find_next_id(info, id, le32_to_cpu(ref[i]), depth + 1);
 		if (ret != -ENOENT)
 			break;
 	}
-- 
2.17.1

